### ISSUE TITLE
How to automatically set actor.eval() when test in training?

### ISSUE THREAD

**Issue by GIS-PuppetMaster** [thu-ml/tianshou#198]
I have some BatchNorm1d Layer with `track_running_stats=True` in my actor net, and I don't want my test set be exposed when test in training.
Q1: Will the `track_running_state` be False is I set model.eval()?
Q2: How to automatically set actor.eval() automatically when test in training and set actor.train() before training without exchanging the source code of Tianshou.TD3policy and offline_trainer?

**Comment by Trinkle23897** [comment]
> Q1: Will the `track_running_state` be False is I set model.eval()?

This question is related to pure PyTorch. The answer is no:

```python
In [10]: import torch

In [11]: m = torch.nn.BatchNorm1d(100, track_running_stats=True)

In [12]: m.training
Out[12]: True

In [13]: m.track_running_stats
Out[13]: True

In [14]: m.eval()
Out[14]: BatchNorm1d(100, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)

In [15]: m.training
Out[15]: False

In [16]: m.track_running_stats
Out[16]: True
```

> Q2: How to automatically set actor.eval() automatically when test in training and set actor.train() before training without exchanging the source code of Tianshou.TD3policy and offline_trainer?

I think it has been done by the current codebase. You can print out the `policy.actor.training` flag to check with it.

**Comment by GIS-PuppetMaster** [comment]
> > Q1: Will the `track_running_state` be False is I set model.eval()?
> 
> This question is related to pure PyTorch. The answer is no:
> 
> ```python
> In [10]: import torch
> 
> In [11]: m = torch.nn.BatchNorm1d(100, track_running_stats=True)
> 
> In [12]: m.training
> Out[12]: True
> 
> In [13]: m.track_running_stats
> Out[13]: True
> 
> In [14]: m.eval()
> Out[14]: BatchNorm1d(100, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
> 
> In [15]: m.training
> Out[15]: False
> 
> In [16]: m.track_running_stats
> Out[16]: True
> ```
> 
> > Q2: How to automatically set actor.eval() automatically when test in training and set actor.train() before training without exchanging the source code of Tianshou.TD3policy and offline_trainer?
> 
> I think it has been done by the current codebase. You can print out the `policy.actor.training` flag to check with it.

Thanks for your patiently answer, and can I use delay reward in Tianshou? For example, each step's reward can only be calculate when the epsiode is done (env.done ==True), then fill the reward back to every (s,a,s',r,info,policy) tuple in Batch().

**Comment by GIS-PuppetMaster** [comment]
So although the current codebase can set actor.eval() and actor.train() automatically, there is no way to set track_running_stats=False/True automatically? Then how should I implement this without changing the source code of Tianshou?

**Comment by Trinkle23897** [comment]
> Can I use delay reward in Tianshou? For example, each step's reward can only be calculate when the episode is done (env.done ==True), then fill the reward back to every (s,a,s',r,info,policy) tuple in Batch().

I think the best way is to add a hook into the `buffer.update` since this function calls when one episode is done and it moves a full-episode data from the cache buffer into the main buffer.
It's very similar to HER. HER implementation will be added soon.
To avoid changing the source code, you can create a buffer class inheriting the `ReplayBuffer`, and overwrite the `update` function.

> So although the current codebase can set actor.eval() and actor.train() automatically, there is no way to set track_running_stats=False/True automatically? Then how should I implement this without changing the source code of Tianshou?

Many solutions: you can create a Policy class inheriting the TD3Policy and overwrite the `train` function, or create a modified bn1d inheriting the `BatchNorm1d` and overwrite its `train`.

**Comment by GIS-PuppetMaster** [comment]
> > Can I use delay reward in Tianshou? For example, each step's reward can only be calculate when the episode is done (env.done ==True), then fill the reward back to every (s,a,s',r,info,policy) tuple in Batch().
> 
> I think the best way is to add a hook into the `buffer.update` since this function calls when one episode is done and it moves a full-episode data from the cache buffer into the main buffer.
> It's very similar to HER. HER implementation will be added soon.
> To avoid changing the source code, you can create a buffer class inheriting the `ReplayBuffer`, and overwrite the `update` function.
> 
> > So although the current codebase can set actor.eval() and actor.train() automatically, there is no way to set track_running_stats=False/True automatically? Then how should I implement this without changing the source code of Tianshou?
> 
> Many solutions: you can create a Policy class inheriting the TD3Policy and overwrite the `train` function, or create a modified bn1d inheriting the `BatchNorm1d` and overwrite its `train`.

Ok, and I just found out there may be a bug in Tianshou, I used BN layer and got '{ValueError}Expected more than 1 value per channel when training, got input size torch.Size([1, 180])', I guess this is because setting actor.train() when collect training sample.

**Comment by GIS-PuppetMaster** [comment]
I know I can fix this by make my customize trainer, but just an advise, maybe add some support to BN layer in the api is easier to use.

**Comment by Trinkle23897** [comment]
> Ok, and I just found out there may be a bug in Tianshou, I used BN layer and got '{ValueError}Expected more than 1 value per channel when training, got input size torch.Size([1, 180])', I guess this is because setting actor.train() when collect training sample.

This issue has been fixed in #185. You can use the current master version. It is because BN doesn't accept batchsize=1.

> I know I can fix this by make my customize trainer, but just an advise, maybe add some support to BN layer in the api is easier to use.

I don't think so. If you are in my point of view, this thing should be fixed in PyTorch instead of tianshou. Moreover, written something as
```python
class xxxPolicy(BasePolicy):
    def train(self):
        # How many actors and critics network, what's their names?
        # Does this issue only occurs in bn1d? How about bn2d and bn3d? How about other PyTorch layers?
        # Does every actor's and critic's bn layer need to set track_running_stats=True/False?
        # Is it reasonable to write this part of BN code here, instead of occurring other confusion?
```
You will see many things to be considered with, it is non-trivial.

**Comment by GIS-PuppetMaster** [comment]
> > Ok, and I just found out there may be a bug in Tianshou, I used BN layer and got '{ValueError}Expected more than 1 value per channel when training, got input size torch.Size([1, 180])', I guess this is because setting actor.train() when collect training sample.
> 
> This issue has been fixed in #185. You can use the current master version. It is because BN doesn't accept batchsize=1.
> 
> > I know I can fix this by make my customize trainer, but just an advise, maybe add some support to BN layer in the api is easier to use.
> 
> I don't think so. If you are in my point of view, this thing should be fixed in PyTorch instead of tianshou. Moreover, written something as
> 
> ```python
> class xxxPolicy(BasePolicy):
>     def train(self):
>         # How many actors and critics network, what's their names?
>         # Does this issue only occurs in bn1d? How about bn2d and bn3d? How about other PyTorch layers?
>         # Does every actor's and critic's bn layer need to set track_running_stats=True/False?
>         # Is it reasonable to write this part of BN code here, instead of occurring other confusion?
> ```
> 
> You will see many things to be considered with, it is non-trivial.

Indeed, thanks for replying anyway.
